[BUG] Writable Warm - Composite Directory sync() method doesn't take into account block files

### Describe the bug

Currently in composite directory, in the [sync() method](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/CompositeDirectory.java#L249) we filter out the files that hare present in Remote and then delegate the sync() call to localDirectory with the filtered set of files. But there can arise cases where the file is present in blocks instead of a full file and that results in a `NoSuchFileException` for the localDirectory sync() call.



Example stacktrace below:

```
2025-06-04T13:21:03,574][ERROR][o.o.i.r.PeerRecoveryTargetService] [c2af93180a503303bf0aa093eb62a955] Exception while downloading segment files from remote store, will continue with peer to peer segment copy
[test_idxx/1010101T7R9GrsTUi9RjDhMmAcPg][[test_idxx][1]] IndexShardRecoveryException[Exception while copying segment files from remote segment store]; nested: NoSuchFileException[/<path>/var/es/data/nodes/0/indices/1010101T7R9GrsTUi9RjDhMmAcPg/1/index/_o_Lucene90_0.dvd];
	at org.opensearch.index.shard.IndexShard.syncSegmentsFromRemoteSegmentStore(IndexShard.java:5294)
	at org.opensearch.indices.recovery.PeerRecoveryTargetService.doRecovery(PeerRecoveryTargetService.java:256)
	at org.opensearch.indices.recovery.PeerRecoveryTargetService$RecoveryRunner.doRun(PeerRecoveryTargetService.java:626)
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1023)
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.nio.file.NoSuchFileException: /<path>/var/es/data/nodes/0/indices/1010101T7R9GrsTUi9RjDhMmAcPg/1/index/_o_Lucene90_0.dvd
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:224)
	at java.base/java.nio.channels.FileChannel.open(FileChannel.java:309)
	at java.base/java.nio.channels.FileChannel.open(FileChannel.java:369)
	at org.apache.lucene.util.IOUtils.fsync(IOUtils.java:465)
	at org.apache.lucene.store.FSDirectory.fsync(FSDirectory.java:303)
	at org.apache.lucene.store.FSDirectory.sync(FSDirectory.java:255)
	at org.opensearch.index.store.CompositeDirectory.sync(CompositeDirectory.java:233)
	at org.apache.lucene.store.FilterDirectory.sync(FilterDirectory.java:86)
	at org.apache.lucene.store.FilterDirectory.sync(FilterDirectory.java:86)
	at org.opensearch.index.store.Store.commitSegmentInfos(Store.java:905)
	at org.opensearch.index.shard.IndexShard.syncSegmentsFromRemoteSegmentStore(IndexShard.java:5290)
	... 7 more
```

### Related component

Storage:Remote

### To Reproduce

Run ingestion for some prolonged time.


### Expected behavior

Composite Directory should be able to handle sync for block files as well.

### Additional Details

N/A

--- a/server/src/main/java/org/opensearch/index/store/CompositeDirectory.java
+++ b/server/src/main/java/org/opensearch/index/store/CompositeDirectory.java
@@ -42,7 +42,6 @@ import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static org.opensearch.index.store.remote.utils.FileTypeUtils.BLOCK_FILE_IDENTIFIER;
-import static org.opensearch.index.store.remote.utils.FileTypeUtils.isBlockFile;
 import static org.apache.lucene.index.IndexFileNames.SEGMENTS;
 
 /**
@@ -103,24 +102,6 @@ public class CompositeDirectory extends FilterDirectory {
     protected List<String> listBlockFiles(String fileName) throws IOException {
         return Stream.of(listLocalFiles())
             .filter(file -> file.equals(fileName) || file.startsWith(fileName + FileTypeUtils.BLOCK_FILE_IDENTIFIER))
-            .collect(Collectors.toList());
-    }
-
-    /**
-     * Returns a list of names of all block files stored in the local directory for a given set of files,
-     * including the original file names itself if present.
-     *
-     * @param fileNames The set of files to search for, along with its associated block files.
-     * @return A list of file names, including the original file (if present) and all its block files.
-     * @throws IOException in case of I/O error while listing files.
-     */
-    protected List<String> listBlockFiles(String[] fileNames) throws IOException {
-        Set<String> files = Set.of(fileNames);
-        return Stream.of(listLocalFiles())
-            .filter(
-                file -> files.contains(file)
-                    || (isBlockFile(file) && files.contains(file.substring(0, file.indexOf(BLOCK_FILE_IDENTIFIER))))
-            )
             .collect(Collectors.toList());
     }
 
@@ -269,16 +250,15 @@ public class CompositeDirectory extends FilterDirectory {
         ensureOpen();
         logger.trace("Composite Directory[{}]: sync() called {}", this::toString, () -> names);
         Set<String> remoteFiles = Set.of(getRemoteFiles());
-        Set<String> localFiles = Arrays.stream(listLocalFiles())
-            .map(file -> isBlockFile(file) ? file.substring(0, file.indexOf(BLOCK_FILE_IDENTIFIER)) : file)
+        Set<String> localFilesHavingBlocks = Arrays.stream(listLocalFiles())
+            .filter(FileTypeUtils::isBlockFile)
+            .map(file -> file.substring(0, file.indexOf(BLOCK_FILE_IDENTIFIER)))
             .collect(Collectors.toSet());
-        String[] fullFilesToSync = names.stream().filter(name -> remoteFiles.contains(name) == false).toArray(String[]::new);
-        for (String fullFileToSync : fullFilesToSync) {
-            if (localFiles.contains(fullFileToSync) == false) throw new NoSuchFileException("Unable to sync file " + fullFileToSync);
-        }
-        List<String> filesToSync = listBlockFiles(fullFilesToSync);
-        logger.trace("Composite Directory[{}]: Synced files : {}", this::toString, () -> filesToSync);
-        localDirectory.sync(filesToSync);
+        Collection<String> fullFilesToSync = names.stream()
+            .filter(name -> (remoteFiles.contains(name) == false) && (localFilesHavingBlocks.contains(name) == false))
+            .collect(Collectors.toList());
+        logger.trace("Composite Directory[{}]: Synced files : {}", this::toString, () -> fullFilesToSync);
+        localDirectory.sync(fullFilesToSync);
     }
 
     /**

--- a/server/src/main/java/org/opensearch/index/store/CompositeDirectory.java
+++ b/server/src/main/java/org/opensearch/index/store/CompositeDirectory.java
@@ -42,6 +42,7 @@ import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static org.opensearch.index.store.remote.utils.FileTypeUtils.BLOCK_FILE_IDENTIFIER;
+import static org.opensearch.index.store.remote.utils.FileTypeUtils.isBlockFile;
 import static org.apache.lucene.index.IndexFileNames.SEGMENTS;
 
 /**
@@ -102,6 +103,24 @@ public class CompositeDirectory extends FilterDirectory {
     protected List<String> listBlockFiles(String fileName) throws IOException {
         return Stream.of(listLocalFiles())
             .filter(file -> file.equals(fileName) || file.startsWith(fileName + FileTypeUtils.BLOCK_FILE_IDENTIFIER))
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Returns a list of names of all block files stored in the local directory for a given set of files,
+     * including the original file names itself if present.
+     *
+     * @param fileNames The set of files to search for, along with its associated block files.
+     * @return A list of file names, including the original file (if present) and all its block files.
+     * @throws IOException in case of I/O error while listing files.
+     */
+    protected List<String> listBlockFiles(String[] fileNames) throws IOException {
+        Set<String> files = Set.of(fileNames);
+        return Stream.of(listLocalFiles())
+            .filter(
+                file -> files.contains(file)
+                    || (isBlockFile(file) && files.contains(file.substring(0, file.indexOf(BLOCK_FILE_IDENTIFIER))))
+            )
             .collect(Collectors.toList());
     }
 
@@ -249,8 +268,15 @@ public class CompositeDirectory extends FilterDirectory {
     public void sync(Collection<String> names) throws IOException {
         ensureOpen();
         logger.trace("Composite Directory[{}]: sync() called {}", this::toString, () -> names);
-        Collection<String> remoteFiles = Arrays.asList(getRemoteFiles());
-        Collection<String> filesToSync = names.stream().filter(name -> remoteFiles.contains(name) == false).collect(Collectors.toList());
+        Set<String> remoteFiles = Set.of(getRemoteFiles());
+        Set<String> localFiles = Arrays.stream(listLocalFiles())
+            .map(file -> isBlockFile(file) ? file.substring(0, file.indexOf(BLOCK_FILE_IDENTIFIER)) : file)
+            .collect(Collectors.toSet());
+        String[] fullFilesToSync = names.stream().filter(name -> remoteFiles.contains(name) == false).toArray(String[]::new);
+        for (String fullFileToSync : fullFilesToSync) {
+            if (localFiles.contains(fullFileToSync) == false) throw new NoSuchFileException("Unable to sync file " + fullFileToSync);
+        }
+        List<String> filesToSync = listBlockFiles(fullFilesToSync);
         logger.trace("Composite Directory[{}]: Synced files : {}", this::toString, () -> filesToSync);
         localDirectory.sync(filesToSync);
     }

--- a/server/src/test/java/org/opensearch/index/store/CompositeDirectoryTests.java
+++ b/server/src/test/java/org/opensearch/index/store/CompositeDirectoryTests.java
@@ -132,6 +132,12 @@ public class CompositeDirectoryTests extends BaseRemoteSegmentStoreDirectoryTest
         // All the files in the below list are present either locally or on remote, so sync should work as expected
         Collection<String> names = List.of("_0.cfe", "_0.cfs", "_0.si", "_1.cfe", "_2.cfe", "segments_1");
         compositeDirectory.sync(names);
+        // Deleting file _1.cfe and then adding its blocks locally so that full file is not present but block files are present in local
+        // State of _1.cfe file after these operations - not present in remote, full file not present locally but blocks present in local
+        compositeDirectory.deleteFile("_1.cfe");
+        addFilesToDirectory(new String[] { "_1.cfe_block_0", "_1.cfe_block_2" });
+        // Sync should work as expected since blocks are present in local
+        compositeDirectory.sync(List.of("_1.cfe"));
         // Below list contains a non-existent file, hence will throw an error
         Collection<String> names1 = List.of("_0.cfe", "_0.cfs", "_0.si", "_1.cfe", "_2.cfe", "segments_1", "non_existent_file");
         assertThrows(NoSuchFileException.class, () -> compositeDirectory.sync(names1));

